### PR TITLE
Fix C# Module Event invocation

### LINKED ISSUE #7797: ReactEvent API changed? Now event data is in array.
I noticed that there is some changes in ReactEvent API on rn-windows 0.64.7. I previously used 0.62.22

If i send event from C# module like this

```
[ReactEvent("folderChangeEvent")]
public Action<JSValue> FolderChangeEvent { get; set; }
```

```
JSValueObject eventData = new JSValueObject() {{ "key", "value") }};
FolderChangeEvent?.Invoke(eventData);
```

Is event value in array of objects in javascript side. There was object in 0.62.x.

```
const myModuleEventEmitter = new NativeEventEmitter(NativeModules.MyCSharpModule);
myModuleEventEmitter.addListener('folderChangeEvent', (event)=>{
    const e = event[0]; // array on 0.64.7
    // const e = event; // object on 0.62.22
});
```

### PR DESCRIPTION
**Pull request by vmoroz**
## Description
Fix C# Module Event invocation

### Type of Change
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Why
Starting with RNW 0.64 we wrap event arguments into an array for the JS events declared as fields or properties in C# modules. It is a bug that prevents using C# native module `[ReactEvent]` fields and properties. Instead, developers must use the `EmitJSEvent` method to work around this issue. Most probably, the bug was due to combining functionality of JS function and event invocation. While they look similar, they require a different way to pass parameters.

Resolves #7797

### What
To resolve the issue we are changing the shape of the code generated by the `Microsoft.ReactNative.Managed.CodeGen` project. Instead of generating a lambda that calls the `WriteArgs`, we are wrapping up the `IReactContext` parameter into the `ReactContext` wrapper class that has `EmitJSEvent` and `CallJSFunction` overloads for multiple arguments that do the correct parameter passing:
  - Use `WriteValue` for the `EmitJSEvent` 
  - Use `WriteArgs` for the `CallJSFunction`

While the bulk of the code generation remains the same, we avoid the wrong use of the `WriteArgs` when calling the `EmitJSEvent`.

Also, we are updating the sample apps with the additional scenarios that helped us to identify and solve the issue.
The work was originally done in 0.65 branch because the sample apps do not work anymore starting with 0.67.

## Testing
The fix was manually tested using the samples apps in the 0.65 branch.
We currently do not have integration tests for C# code.
These tests must be added as a part of a different PR.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/9154)

### CODE REVIEW

**Review by NikoAri — APPROVED** [review]
:shipit:

**Comment by NickGerleman** [comment]
> We currently do not have integration tests for C# code.

We do have tests against CS modules from a running JS instance. See `packages\integration-test-app\tests\InteropCSModuleTests.tsx`

**Comment by vmoroz** [comment]
Good to know! Unfortuntely, it is not clear how to use/change/run it - the ReadMe.md is empty.
I am looking for creating something like Microsoft.ReactNative.IntegrationTests that we have for C++ code.
It would be great to reuse the same JS files between C++ and C# tests.

---
In reply to: [970810085](https://github.com/microsoft/react-native-windows/pull/9154#issuecomment-970810085) [](http://example.com/codeflow?ancestors=970810085,970810085)

**Review by dannyvv — APPROVED** [review]

**Comment by NickGerleman** [comment]
Oh yes, I never got around to filling out the readme...

Test authoring should be relatively straightforward from the JS files, and tests can be run by launching the app, then running `yarn integration-test` from the `integration-test-app` directory. There is currently a bug where the app may need to be restarted to run additional tests from the command line.

**Comment by vmoroz** [comment]
I will have a look, but not as a part of this PR.

---
In reply to: [970849650](https://github.com/microsoft/react-native-windows/pull/9154#issuecomment-970849650) [](http://example.com/codeflow?ancestors=970849650,970849650)

File: vnext/Microsoft.ReactNative.Managed.CodeGen/SyntaxHelpers.cs [hunk 1]
```diff
@@ -38,6 +38,14 @@ internal static ExpressionStatementSyntax InvocationStatement(SyntaxToken method
       return InvocationStatement(SyntaxFactory.IdentifierName(methodName), arguments);
     }
 
+    internal static ExpressionSyntax MemberAccessExpression(ExpressionSyntax instance, SyntaxToken method)
```
Comment (asklar) [inline vnext/Microsoft.ReactNative.Managed.CodeGen/SyntaxHelpers.cs:41]: nit, rename method to member?

**Review by asklar — APPROVED** [review]

File: vnext/Microsoft.ReactNative.Managed.CodeGen/SyntaxHelpers.cs (hunk 1 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Managed.CodeGen/SyntaxHelpers.cs:41]: good point!

**Comment by NickGerleman** [comment]
@dannyvv are there docs anywhere on how to update the lkgs?

**Comment by vmoroz** [comment]
> @dannyvv are there docs anywhere on how to update the lkgs?

I was able to find the variable in C# code that controls it while debugging the failing test. See the AutoFixLkgs variable in the vnext\Microsoft.ReactNative.Managed.CodeGen.UnitTests\CodeGenTestBase.cs file.

**Pull request merged**